### PR TITLE
`Client.net` deprecation

### PR DESCRIPTION
**Pull request by war-in**
Closes #

## Introduced changes

- Client.net is deprecated
- Account.get_balance has chain_id parameter
- net parameter in FullNodeClient's constructor is deprecated

##

- [x] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/net/account/account.py [hunk 1]
```diff
@@ -96,11 +86,15 @@ def __init__(
...
                 )
 
-            chain = chain_from_network(net=client.net, chain=chain)
             signer = StarkCurveSigner(
-                account_address=self.address, key_pair=key_pair, chain_id=chain
+                account_address=self.address,
+                key_pair=key_pair,
+                chain_id=cast(
+                    StarknetChainId, chain
+                ),  # cast is required, because pyright doesn't know chain is not None here
```
Comment (drknzz) [inline starknet_py/net/account/account.py:94]: How come it wasn't required before?

File: starknet_py/net/models/chains.py [hunk 2]
```diff
@@ -27,3 +28,16 @@ def chain_from_network(
...
+
+
+def default_token_address_for_chain(chain_id: Optional[StarknetChainId] = None) -> str:
+    if chain_id not in [
+        StarknetChainId.TESTNET,
+        StarknetChainId.TESTNET2,
+        StarknetChainId.MAINNET,
+    ]:
+        raise ValueError(
+            "Argument token_address must be specified when using a custom net."
```
Comment (drknzz) [inline starknet_py/net/models/chains.py:40]: Should it still be net?

File: starknet_py/net/account/account.py [hunk 3]
```diff
@@ -224,10 +218,12 @@ async def get_nonce(self) -> int:
...
     async def get_balance(
-        self, token_address: Optional[AddressRepresentation] = None
+        self,
+        token_address: Optional[AddressRepresentation] = None,
+        chain_id: Optional[StarknetChainId] = None,
     ) -> int:
-        token_address = token_address or default_token_address_for_network(
-            self._client.net
+        token_address = token_address or default_token_address_for_chain(
+            chain_id or self._chain_id
```
Comment (drknzz) [inline starknet_py/net/account/account.py:226]: Could we maybe do something more explicit like
```
if token_address is None:
   chain_id = chain_id or self._chain_id
   token_address = default_token_address_for_chain(chain_id)
```

File: starknet_py/tests/e2e/utils.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ async def get_deploy_account_details(
     )
 
     res = await fee_contract.functions["transfer"].invoke(
-        recipient=address, amount=int(1e15), max_fee=MAX_FEE
+        recipient=address, amount=int(1e20), max_fee=MAX_FEE
```
Comment (drknzz) [inline starknet_py/tests/e2e/utils.py:40]: Why this change?

File: starknet_py/net/account/account.py (hunk 1 above)
Comment (war-in) [reply starknet_py/net/account/account.py:94]: `chain_from_network()` returns StarknetChainId or raises an error, so the chain was always StarknetChainId here

File: starknet_py/net/models/chains.py (hunk 2 above)
Comment (war-in) [reply starknet_py/net/models/chains.py:40]: Maybe `network`

File: starknet_py/net/account/account.py (hunk 1 above)
Comment (drknzz) [reply starknet_py/net/account/account.py:94]: We could try doing an `assert` 👀

Comment (drknzz) [reply starknet_py/net/account/account.py:94]: Casting in situations like these always requires an additional comment explaining the reason behind it.
I think
```
assert chain is not None
```
is shorter and self explanatory

Comment (war-in) [reply starknet_py/net/account/account.py:94]: I know who would not like it 🙈

Comment (drknzz) [reply starknet_py/net/account/account.py:94]: 😆
It wasn't about the use of `assert` though but rather the same `assert` in multiple places in code, implying a flaw in the structure of the code

Comment (war-in) [reply starknet_py/net/account/account.py:94]: I'll leave it as open discussion 😅

Comment (Arcticae) [reply starknet_py/net/account/account.py:94]: It is perfectly viable for `chain` to be `None` here, so i don't know why the `cast` is here - py**right** seems to be **right** 🙃

File: starknet_py/net/account/base_account.py [hunk 5]
```diff
@@ -51,13 +51,16 @@ async def get_nonce(self) -> int:
...
+        token_address: Optional[AddressRepresentation] = None,
+        chain_id: Optional[StarknetChainId] = None,
     ) -> int:
         """
         Checks account's balance of specified token.
 
         :param token_address: Address of the ERC20 contract.
-            If not specified it will be payment token address.
+        :param chain_id: Identifier of the Starknet chain used.
+            If token_address is not specified it will be used to determine network's payment token address.
```
Comment (Arcticae) [inline starknet_py/net/account/base_account.py:63]: it would be also irrelevant when `token_address` is provided - worth to note here

File: starknet_py/net/models/chains.py [hunk 6]
```diff
@@ -27,3 +28,16 @@ def chain_from_network(
...
     return chain
+
+
+def default_token_address_for_chain(chain_id: Optional[StarknetChainId] = None) -> str:
+    if chain_id not in [
+        StarknetChainId.TESTNET,
+        StarknetChainId.TESTNET2,
+        StarknetChainId.MAINNET,
+    ]:
+        raise ValueError(
```
Comment (Arcticae) [inline starknet_py/net/models/chains.py:39]: [nitpick]: The error message here might depend on the function usage context, so i'd avoid throwing in those kind of methods. I guess for the use-case is fine, just noting here.

File: starknet_py/net/models/chains.py [hunk 7]
```diff
@@ -27,3 +28,16 @@ def chain_from_network(
         raise ValueError("Chain is required when not using predefined networks.")
 
     return chain
+
+
+def default_token_address_for_chain(chain_id: Optional[StarknetChainId] = None) -> str:
```
Comment (Arcticae) [inline starknet_py/net/models/chains.py:33]: i struggle to understand why the chain_id can be `None` here, kind of weird function definition.

File: starknet_py/net/full_node_client.py [hunk 8]
```diff
@@ -65,10 +66,14 @@ def __init__(
...
         self._client = RpcHttpClient(url=node_url, session=session)
+
+        if net is not None:
+            warnings.warn("Parameter net is deprecated.", category=DeprecationWarning)
         self._net = net
 
     @property
-    def net(self) -> Network:
+    def net(self) -> Optional[Network]:
+        warnings.warn("Parameter net is deprecated.", category=DeprecationWarning)
```
Comment (Arcticae) [inline starknet_py/net/full_node_client.py:76]: This doesn't tell which class doesn't accept the parameter anymore. Did you check manually how does the trace look when using the library with `net`, and if user can tell where the `warn` has its' origin?

**Review by Arcticae — COMMENTED** [review]
It looks good from API-construction perspective, i'd circle back on the `cast` though.
Also, some nitpicking but overall looks fine 👍

File: starknet_py/net/models/chains.py (hunk 7 above)
Comment (war-in) [reply starknet_py/net/models/chains.py:33]: It can be None, because 
1. an account can be created without passing the chain
2. if 1 and the user does not provide chain_id in the get_balance => we do not know the chain

File: starknet_py/net/account/account.py (hunk 1 above)
Comment (Arcticae) [reply starknet_py/net/account/account.py:94]: I actually think that having this kind of implicit logic here makes for confusing code reading experience, did you guys consider changing it?

File: starknet_py/net/models/chains.py (hunk 7 above)
Comment (Arcticae) [reply starknet_py/net/models/chains.py:33]: Can't `StarknetChainId` have this method?

File: starknet_py/net/account/account.py (hunk 1 above)
Comment (drknzz) [reply starknet_py/net/account/account.py:94]: **chain** cannot be `None` here, the check for it is few lines above:
```
        if chain is None and signer is None:
            raise ValueError("One of chain or signer must be provided.")
```

Comment (drknzz) [reply starknet_py/net/account/account.py:94]: I agree that the code is hard to follow though.
There was some attempt at making it more explicit in [this _closed_ PR](https://github.com/software-mansion/starknet.py/pull/736/files#diff-71fc6b5ba6c214b0ddf79941772b54fef34507113a848e666dbc8e9c1002a0f3).
Maybe it should be implemented in some other PR nonetheless.

Comment (war-in) [reply starknet_py/net/account/account.py]: it is fixed now! Check it out

**Review by drknzz — APPROVED** [review]

**Review by Arcticae — APPROVED** [review]
